index.d.ts

```diff
@@ -1,148 +1,345 @@
-}
-
-interface NumberPromptOptions extends BasePromptOptions {
-  type: 'numeral'
-  min?: number
-  max?: number
-  delay?: number
-  float?: boolean
-  round?: boolean
-  major?: number
-  minor?: number
-  initial?: number
-}
-
-interface SnippetPromptOptions extends BasePromptOptions {
-  type: 'snippet'
-  newline?: string
-}
-
-interface SortPromptOptions extends BasePromptOptions {
-  type: 'sort'
-  hint?: string
-  drag?: boolean
-  numbered?: boolean
-}
+export = EnquirerStatic;
 
-type PromptOptions =
-  | ArrayPromptOptions
-  | BooleanPromptOptions
-  | StringPromptOptions
-  | NumberPromptOptions
-  | SnippetPromptOptions
-  | SortPromptOptions
-  | BasePromptOptions
-
-declare class BasePrompt extends NodeJS.EventEmitter {
-    constructor(options?: PromptOptions);
-
-    render(): void;
-
-    run(): Promise<any>;
-  }
-
-declare class Enquirer<T = object> extends NodeJS.EventEmitter {
-  constructor(options?: object, answers?: T);
+declare class EnquirerStatic<
+  R = any,
+  N extends string = string
+> extends NodeJS.EventEmitter {
+  // TODO: update object options
+  constructor(options?: object, answers?: object);
 
   /**
    * Register a custom prompt type.
    *
    * @param type
    * @param fn `Prompt` class, or a function that returns a `Prompt` class.
    */
-  register(type: string, fn: typeof BasePrompt | (() => typeof BasePrompt)): this;
+  register(type: string): this;
 
   /**
-   * Register a custom prompt type.
+   * Prompt function that takes a "question" object or array of question objects,
+   * and returns an object with responses from the user.
+   *
+   * @param question Array of question for prompts to run.
    */
-  register(type: { [key: string]: typeof BasePrompt | (() => typeof BasePrompt) }): this;
+  prompt(
+    questions: Array<EnquirerStatic.PromptOptions<N>>
+  ): Promise<EnquirerStatic.PromptType.Answers<N, R>>;
 
   /**
    * Prompt function that takes a "question" object or array of question objects,
    * and returns an object with responses from the user.
    *
-   * @param questions Options objects for one or more prompts to run.
+   * @param question Question for prompts to run.
    */
-  prompt(
-    questions:
-      | PromptOptions
-      | ((this: Enquirer) => PromptOptions)
-      | (PromptOptions | ((this: Enquirer) => PromptOptions))[]
-  ): Promise<T>;
+  prompt(question: EnquirerStatic.PromptOptions<N>): Promise<R>;
 
   /**
    * Use an enquirer plugin.
    *
    * @param plugin Plugin function that takes an instance of Enquirer.
    */
-  use(plugin: (this: this, enquirer: this) => void): this;
+  use(plugin: EnquirerStatic.Caller.BindOneParam<this, this, void>): this;
 }
 
-declare namespace Enquirer {
-  function prompt<T = object>(
-    questions:
-      | PromptOptions
-      | ((this: Enquirer) => PromptOptions)
-      | (PromptOptions | ((this: Enquirer) => PromptOptions))[]
-  ): Promise<T>;
+declare namespace EnquirerStatic {
+  namespace Caller {
```

Comment:
Maybe we can merge `Caller` with `UnionType` namespace as a new namespace called `TypeUtils`?

Reply:
yes.